File: 2013/cracking/01-01.py
# coding=utf8
# Implement an algorithm to determine if a string has all unique characters. What if you can not use additional data structures?
def check_for_unique(value):
    "Determine if string has all unique characters"
    count = [False] * 256
    for c in value:
        index = ord(c) 

        if count[index] == True:
            return False
            
        count[index] = True
        
    return True;
        
print "This string 'Hello' is unique: %s" % check_for_unique("Hello")
print "This string 'Helo' is unique: %s" % check_for_unique("Helo")

File: 2013/cracking/01-01.c
// Implement an algorithm to determine if a string has all unique characters. 
// What if you can not use additional data structures?

#include <stdio.h>
#include <stdbool.h>

bool is_unique(char* str)
{
  bool a[256] = {0};

  char* cursor = str;
  while (*cursor != '\0')
  {
    int charIndex = *cursor;

    if (a[charIndex])
    {
      return false;
    }

    a[charIndex] = true;

    cursor++;
  }

  return true;
}

int main()
{
  char test[] = "This";
  printf("This string has only unique values: %d\n", is_unique(test));
  return 0;
}